include/sycl/algorithm/inclusive_scan.hpp

```diff
@@ -107,7 +109,256 @@ OutputIterator inclusive_scan(ExecutionPolicy &sep, InputIterator b,
+    size(size_),
+    size_per_work_group(size_per_work_group_),
+    size_per_work_item(size_per_work_item_),
+    nb_work_group(nb_work_group_),
+    nb_work_item(nb_work_item_) {}
+};
+
+mapscan_descriptor compute_mapscan_descriptor(cl::sycl::device device,
+                                              size_t size,
+                                              size_t sizeofB) {
+  //std::cout << "size=\t" << size << std::endl;
+  using std::min;
+  using std::max;
+  if (size == 0)
+    return mapscan_descriptor(0, 0, 0, 0, 0);
+  size_t local_mem_size =
+    device.get_info<cl::sycl::info::device::local_mem_size>();
+  //std::cout << "local_mem_size=\t" << local_mem_size << std::endl;
+  size_t size_per_work_group = min(size, local_mem_size / sizeofB);
+  //std::cout << "size_per_work_group=\t" << size_per_work_group << std::endl;
+  if (size_per_work_group <= 0)
+    return mapscan_descriptor(size, 0, 0, 0, 0);
+
+  size_t nb_work_group = up_rounded_division(size, size_per_work_group);
+  //std::cout << "nb_work_group=\t" << nb_work_group << std::endl;
+
+  size_t max_work_item =
+    device.get_info<cl::sycl::info::device::max_work_group_size>();
+  //std::cout << "max_work_item=\t" << max_work_item << std::endl;
+  size_t nb_work_item = min(max_work_item, size_per_work_group);
+  //std::cout << "nb_work_item=\t" << nb_work_item << std::endl;
+  size_t size_per_work_item =
+    up_rounded_division(size_per_work_group, nb_work_item);
+  //std::cout << "size_per_work_item=\t" << size_per_work_item << std::endl;
+  return mapscan_descriptor(size, size_per_work_group,
+                            size_per_work_item, nb_work_group, nb_work_item);
+}
+
+
+template <class ExecutionPolicy, class A, class B, class Reduce, class Map>
+void buffer_mapscan(ExecutionPolicy &snp,
+                    cl::sycl::queue q,
+                    cl::sycl::buffer<A, 1> input_buffer,
+                    cl::sycl::buffer<B, 1> output_buffer,
+                    B init,
+                    mapscan_descriptor d,
+                    Map map,
+                    Reduce red) {
+    //map is not applied on init
+
+  using std::min;
+  using std::max;
+
+  size_t size = d.size;
+  size_t size_per_work_group = d.size_per_work_group;
+  size_t size_per_work_item = d.size_per_work_item;
+  size_t nb_work_group = d.nb_work_group;
+  size_t nb_work_item = d.nb_work_item;
+
+  //WARNING: nb_work_group is not bounded by max_compute_units
+  cl::sycl::buffer<B, 1> scan = { cl::sycl::range<1> { nb_work_group } };
+
+  q.submit([&] (cl::sycl::handler &cgh) {
+    cl::sycl::nd_range<1> rng
+      { cl::sycl::range<1>{nb_work_group * nb_work_item},
+        cl::sycl::range<1>{nb_work_item} };
+    auto input =
+      input_buffer.template get_access<cl::sycl::access::mode::read>(cgh);
+    auto output =
+      output_buffer.template get_access<cl::sycl::access::mode::write>(cgh);
+
+    cgh.parallel_for_work_group<class wg>(rng, [=](cl::sycl::group<1> grp) {
```

Comment:
Ah yes. These are triSYCL-ism.
For example we do not use the template type name of `parallel_for_work_group` for now.